### PR TITLE
Som devel 20 refac

### PR DESCRIPTION
**Pull request by sorgom**

### CODE REVIEW

File: application/components/BAS/BAS_Provider.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
+//  ============================================================
+//  class BAS_Provider: base class to implement I_Provider
+//  - provides everything to satisfy 
+//    an I_Provider derived interface
```
Comment (sombeam) [inline application/components/BAS/BAS_Provider.h:4]: a interface

File: application/components/BAS/BAS_Provider.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+//  - except the load method which is too specific
```
Comment (sombeam) [inline application/components/BAS/BAS_Provider.h]: it's not the load method, it's the add and comp method

File: application/components/BAS/BAS_Provider.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        mElems.clear();
+    }
+
+    void load(const ProjItem* data, UINT32 num) final;
+    
+protected:
+    inline BAS_Provider() = default;
+
+    virtual bool add(size_t id, const ProjItem& item) = 0;
```
Comment (sombeam) [inline application/components/BAS/BAS_Provider.h:42]: comment methods

File: specification/ifs/DataTypes.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+};
+
+static_assert(ComNameSize == sizeof(ComName));
+
+//  standard telegram size
+constexpr auto ComTelegramSize = ComNameSize + 2;
+
+//  Com telegrams data
+struct ComData
```
Comment (sombeam) [inline specification/ifs/DataTypes.h:55]: maybe add a constructor with PARAM_UNDEF as default

File: application/components/SIG/SIG_X.h [hunk 5]
```diff
@@ -30,6 +32,15 @@ class SIG_X : public I_SIG
     void procFromFld(UINT8 state);
     void procFromGui(UINT8 stateFld, UINT8 stateGui);
 
+    inline void toFld(UINT8 state, UINT8 speed = PARAM_UNDEF)
```
Comment (sombeam) [inline application/components/SIG/SIG_X.h]: inline is not gcov testable

File: application/components/SYS/src/Dispatcher.cpp [hunk 6]
```diff
@@ -85,28 +85,51 @@ void Dispatcher::dispatch(const ComGuiCmd& tele) const
...
 }
 
-void Dispatcher::dispatch(const size_t id, ComCmdFld&& tele) const
+void Dispatcher::toFld(const size_t id, const ComData& data) const
 {
     if (mIndx.size() > id)
     {
-        tele.name = mIndx.at(id).name;
-        IL::getCom().send(tele);
+        const ComTele tele { mIndx.at(id).name, data };
```
Comment (sombeam) [inline application/components/SYS/src/Dispatcher.cpp:96]: could be done without telegram object

File: application/components/SYS/src/Dispatcher.cpp [hunk 7]
```diff
@@ -85,28 +85,51 @@ void Dispatcher::dispatch(const ComGuiCmd& tele) const
...
 }
 
-void Dispatcher::dispatch(const size_t id, ComStateGui&& tele) const
+void Dispatcher::toGui(const size_t id, const ComData& data) const
 {
     if (mIndx.size() > id)
     {
-        tele.name = mIndx.at(id).name;
-        IL::getCom().send(tele);
+        const ComTele tele { mIndx.at(id).name, data };
```
Comment (sombeam) [inline application/components/SYS/src/Dispatcher.cpp:107]: could be done without telegram object

File: application/components/SYS/src/Reader.cpp [hunk 8]
```diff
@@ -39,27 +39,28 @@ void Reader::read(const CONST_C_STRING filename) const
             is.read(head.buf, hSize);
             auto [nTSW, nSIG, nLCR, nSEG] = head.vals;
 
-            const stype sTSW = nTSW * sizeof(ProjTSW);
-            const stype sSIG = nSIG * sizeof(ProjSIG);
-            const stype sLCR = nLCR * sizeof(ProjLCR);
-            const stype sSEG = nSEG * sizeof(ProjSEG);
+            const stype sTSW = nTSW * sizeof(ProjItem);
```
Comment (sombeam) [inline application/components/SYS/src/Reader.cpp:42]: maybe make a const expr of the size

File: application/components/SYS/src/Reader.cpp [hunk 9]
```diff
@@ -39,27 +39,28 @@ void Reader::read(const CONST_C_STRING filename) const
...
+            ok = (sTOT > 0) and (fsize == hSize + sTOT);
 
             if (ok)
             {
+                
                 const auto mxSize = std::max({sTSW, sSIG, sLCR, sSEG});
-
                 CHAR* buf = new CHAR[static_cast<size_t>(mxSize)];
 
                 is.read(buf, sTSW);
```
Comment (sombeam) [inline application/components/SYS/src/Reader.cpp:56]: maybe comment that read 0 doesn't do any harm

File: devel/TODO.md
Comment (sombeam) [inline devel/TODO.md:1]: maybe remove the whole file: use issues

File: make/build.sh [hunk 10]
```diff
@@ -48,15 +48,23 @@ if test ! -z $clean; then git clean -dfXq .; fi
 if test ! -z $pre; then premake5 gmake2; fi
 
 echo build ...
+err=0
 pids=()
-for fn in $(ls *.make | grep -v '_'); do
+for fn in lib_*.make; do
```
Comment (sombeam) [inline make/build.sh:53]: just to remember: could we build some more libs here? e.g. coverage application

File: make/premake5_vs.lua
Comment (sombeam) [inline make/premake5_vs.lua:1]: maybe add system tests

File: scripts/genCode.py
Comment (sombeam) [inline scripts/genCode.py:1]: remove the whole genCode stuff

File: scripts/modScanCode.py
Comment (sombeam) [inline scripts/modScanCode.py:1]: remove the code gen stuff

File: scripts/modTransTable.py [hunk 11]
```diff
@@ -132,7 +132,7 @@ def cppName(self, prefix:str, state:str):
     def genCpp(self, transEvents:list, 
             prefixState = 'SIG_STATE',
             prefixCmd = 'SIG_STATE',
-            cmd1 = 'CMD(_VAL_, 0, _VAL_, 0, _TRG_, 0);',
+            param1 = 'CMD(_VAL_, 0, _VAL_, 0, _TRG_, 0);',
```
Comment (sombeam) [inline scripts/modTransTable.py]: hm, why is that param1 an not cmd1

File: specification/ifs/I_Provider.h
Comment (sombeam) [inline specification/ifs/I_Provider.h:1]: in general: describe the methods

File: specification/ifs/values.h [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+//  ============================================================
+enum E_Param : UINT8
+{
+    PARAM_UNDEF = 255
+};
+
+//  ============================================================
+//  - SIG types and states
+//  ============================================================
+enum E_SigType : UINT8
```
Comment (sombeam) [inline specification/ifs/values.h]: maybe spend a common types enum instead

File: testing/testenv/comparators/ostreams.h [hunk 13]
```diff
@@ -9,25 +9,14 @@
...
 OSTREAM_DEC(ComName)
-OSTREAM_DEC(ComStateGui)
+OSTREAM_DEC(ComTele)
 OSTREAM_DEC(PosRes)
-OSTREAM_DEC(ProjLCR)
-OSTREAM_DEC(ProjSEG)
-OSTREAM_DEC(ProjSIG)
-OSTREAM_DEC(ProjTSW)
-//# END
+OSTREAM_DEC(ProjItem)
```
Comment (sombeam) [inline testing/testenv/comparators/ostreams.h:20]: is that defined and used?

File: testing/testenv/comparators/src/installComparators.cpp [hunk 14]
```diff
@@ -1,24 +1,21 @@
...
 #define ADD_COMP(NAME) \
     static Comparator<NAME> cmp_ ## NAME; \
-    plugin.installComparator(c__ ## NAME, cmp_ ## NAME);
+    plugin.installComparator(#NAME, cmp_ ## NAME);
 
 namespace test
 {
     void installComparators(MockSupportPlugin& plugin)
     {
         //# ADD_COMP
```
Comment (sombeam) [inline testing/testenv/comparators/src/installComparators.cpp]: remove the gen code tags generally

File: testing/testenv/mocks/M_Elem.h [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    class M_LCR : public M_Elem
+    {
+    public:
+        inline M_LCR() : M_Elem("LCR") {}
+        INSTANCE_DEC(M_LCR)
+    };
+    
+} // namespace
+
```
Comment (sombeam) [inline testing/testenv/mocks/M_Elem.h:66]: remove empty lines

File: testing/testenv/mocks/M_Instances.h [hunk 16]
```diff
@@ -10,46 +10,29 @@
...
-//## END
-
-#define MOCK_DEC(NAME) \
-    inline M_ ## NAME& m_ ## NAME() { return M_ ## NAME::instance(); } \
-    inline void mock_ ## NAME() { IL::set ## NAME(m_ ## NAME()); }
+#include "M_Provider.h"
 
 namespace test
 {
-    //# MOCK_DEC
```
Comment (sombeam) [inline testing/testenv/mocks/M_Instances.h:36]: remove gen code tags

File: testing/testenv/mocks/src/mock_IL.cpp [hunk 17]
```diff
@@ -0,0 +1,15 @@
+#include <SYS/IL.h>
+
+#include <mocks/M_Instances.h>
+#include <BAS/coding.h>
```
Comment (sombeam) [inline testing/testenv/mocks/src/mock_IL.cpp]: what for?

File: testing/tests/devtests/DT_01.cpp
Comment (sombeam) [inline testing/tests/devtests/DT_01.cpp:1]: make clear what this is for or remove

File: testing/tests/moduletests/SYS/SYS_01.cpp [hunk 18]
```diff
@@ -12,17 +12,17 @@ namespace test
...
-            m_Dispatcher().expectReset();
-            m_TSW_Provider().expectReset();
-            m_SIG_Provider().expectReset();
-            m_LCR_Provider().expectReset();
+            m_Dispatcher().expectClear();
+            m_TSW_Provider().expectClear();
+            m_SIG_Provider().expectClear();
+            m_LCR_Provider().expectClear();
         }
         void expectFail()
```
Comment (sombeam) [inline testing/tests/moduletests/SYS/SYS_01.cpp:22]: is this used?

**Review by sombeam — CHANGES_REQUESTED** [review]

File: application/components/BAS/BAS_Provider.h (hunk 1 above)
Comment (sorgom) [reply application/components/BAS/BAS_Provider.h:4]: an I_Provider...

File: testing/testenv/mocks/M_Instances.h (hunk 16 above)
Comment (sorgom) [reply testing/testenv/mocks/M_Instances.h:36]: this was already removed

File: testing/tests/moduletests/SYS/SYS_01.cpp (hunk 18 above)
Comment (sorgom) [reply testing/tests/moduletests/SYS/SYS_01.cpp:22]: yes twice

File: testing/tests/devtests/DT_01.cpp
Comment (sorgom) [reply testing/tests/devtests/DT_01.cpp:1]: we'll leave that for now

File: specification/ifs/DataTypes.h (hunk 4 above)
Comment (sorgom) [reply specification/ifs/DataTypes.h:55]: added default to members, seems to work

**Review by sombeam — APPROVED** [review]

**Pull request merged**